bin/build.sh

```diff
@@ -7,118 +7,92 @@ cd "$(dirname "${BASH_SOURCE[0]}")/.."
+		'containerd' snapshotter, and as a result will not work in CI
+		environments.
+
+		Examples:
+		${BASE_NAME} 24
+		${BASE_NAME} 24 arm64 amd64
 	EOF
 }
 
 [[ $STACK_VERSION =~ ^[0-9]+$ ]] || (>&2 print_usage && abort "fatal: invalid STACK_VERSION")
 
-have_docker_container_driver=
-if (docker buildx inspect; true) | grep -q 'Driver:\s*docker-container$'; then
-	have_docker_container_driver=1
-fi
-
 have_containerd_snapshotter=
-if docker info -f "{{ .DriverStatus }}" | grep -qF "io.containerd.snapshotter."; then have_containerd_snapshotter=1; fi
-
+if docker info -f "{{ .DriverStatus }}" | grep -qF "io.containerd.snapshotter."; then
+	have_containerd_snapshotter=1;
+fi
 
 if (( STACK_VERSION <= 22 )); then
 	# heroku/heroku:22 and prior images do not support multiple chip
 	# architectures or multi-arch images. Instead, they are amd64 only.
-	DOCKER_ARGS=("build" "--platform=linux/amd64")
+	if (( ${#ARCHS[@]} == 0 )); then
+		ARCHS=("amd64")
+	fi
 	# heroku/heroku:22 and prior images need separate *cnb* variants that
 	# add compatibility for Cloud Native Buildpacks.
 	VARIANTS=("-build:" "-cnb:" "-cnb-build:-build")
 else
-	# heroku/heroku:24 images and beyond are multi-arch (amd64+arm64) images.
-	# Due to weak feature support parity between Docker on Linux and Docker
-	# Desktop building and publishing across platforms has caveats (see the
-	# top of this file).
-	if [[ $have_containerd_snapshotter ]] || { [[ $PUBLISH_SUFFIX ]] && [[ $have_docker_container_driver ]]; }; then
-		DOCKER_ARGS=("buildx" "build" "--platform=linux/amd64,linux/arm64")
-	elif [[ ! $PUBLISH_SUFFIX ]] && [[ ! $have_docker_container_driver ]]; then
-		DOCKER_ARGS=("buildx" "build" "--platform=linux/amd64")
-		>&2 echo "WARNING: heroku-24 and newer images are multi-arch images," \
-			"but this script is building single architecture images" \
-			"due to limitations of the current platform." \
-			"To build a multi-arch image, enable the 'containerd'" \
-			"snapshotter in Docker Desktop and/or use a 'docker-container'" \
-			"Docker BuildKit driver."
-	else
-		>&2 echo "ERROR: Can't build images with this configuration. Enable" \
-			"the 'containerd' snapshotter in Docker Desktop, enable" \
-			"the 'docker-container' driver in Docker, or use this script" \
-			"in build-only mode (don't provide PUBLISH_SUFFIX argument)."
-		abort 1
+	# heroku/heroku:24 and beyond support amd64 and arm64 and are published
+	# as multi-arch images.
+	if (( ${#ARCHS[@]} == 0 )); then
+		ARCHS=("amd64" "arm64")
 	fi
 	# heroku/heroku:24 and beyond images include CNB specific
 	# modifications, so separate *cnb* variants are not created.
 	VARIANTS=("-build:")
 fi
 
-if [[ $PUBLISH_SUFFIX ]]; then
-	# If there is a tag suffix, this script is pushing to a remote registry.
-	DOCKER_ARGS+=("--push")
-else
-	# Otherwise, load the image into the local image store.
-	DOCKER_ARGS+=("--load")
+if (( ${#ARCHS[@]} > 1 )) && (( STACK_VERSION <= 22 )) ; then
```

Comment:
(optional) This conditional permits trying to build ARM64 images for older stacks - eg `./bin/build.sh 22 arm64` works locally. I guess in practice this doesn't matter much - but then it we don't mind attempting to build arm64 on older stacks, do we even need this `Can't build multi-arch images for heroku-22 and prior` error case at all? Perhaps either this conditional should be made more robust or else removed entirely?

(We obviously still need the containerd snapshotter check in the lines following)

Reply:
Yeah, fair enough. If we allow heroku-22 arm64 builds, we might as well allow heroku-22 arm64+amd64 builds too. I've removed this.